### PR TITLE
docs(audit): document thread-safety guarantees for global PluginRegistry (#290)

### LINKED ISSUE #290: Document thread-safety guarantees for global PluginRegistry access pattern
## Summary

`internal/audit/plugin.go:260-277` uses `sync.Once` to initialize the global registry. While this pattern is technically safe under Go's memory model, the code lacks documentation clarifying the thread-safety guarantees and usage constraints for concurrent access.

## Current Implementation

```go
var (
    globalRegistry     *PluginRegistry
    globalRegistryOnce sync.Once
)

func GetGlobalRegistry() *PluginRegistry {
    globalRegistryOnce.Do(func() {
        globalRegistry = NewPluginRegistry()
    })
    return globalRegistry  // No explicit documentation of memory barriers
}

func RegisterGlobalPlugin(p compliance.Plugin) error {
    return GetGlobalRegistry().RegisterPlugin(p)
}
```

### Go Memory Model Guarantees
According to the [Go memory model](https://go.dev/ref/mem#once), `sync.Once.Do(f)` provides the following guarantee:
> A single call to f() from once.Do(f) happens (returns) before any call of once.Do(f) returns.

This means:
- ✅ The initialization inside `Do()` happens-before any subsequent `GetGlobalRegistry()` call returns
- ✅ All writes to `globalRegistry` are visible to subsequent readers
- ✅ Multiple concurrent calls to `GetGlobalRegistry()` are safe

### Current Usage Pattern
The internal `PluginRegistry` methods properly use `sync.RWMutex` for plugin map access:
- `RegisterPlugin()` uses `mutex.Lock()`
- `GetPlugin()` and `ListPlugins()` use `mutex.RLock()`

### Actual Risk Assessment
**Risk Level:** Very Low

The current code is **technically correct** under Go's memory model. However:
- ⚠️ The thread-safety guarantees are implicit and undocumented
- ⚠️ Code reviewers may flag this pattern as potentially unsafe
- ⚠️ Future maintainers may introduce bugs by misunderstanding the pattern
- ⚠️ The initialization-before-use invariant is not explicitly documented

## Recommended Solution

**Option A: Documentation (Recommended)**

Add clear documentation about thread-safety guarantees and usage constraints:

```go
// globalRegistry holds the singleton plugin registry instance.
// Thread-safety: GetGlobalRegistry() uses sync.Once to ensure initialization
// happens exactly once. The Go memory model guarantees that all writes within
// sync.Once.Do() are visible to any goroutine that observes the completion of Do().
// 
// Usage constraint: All plugin registration via RegisterGlobalPlugin() should
// complete during application initialization (InitializePlugins()) before
// concurrent access begins. While GetGlobalRegistry() is safe to call concurrently,
// the current architecture assumes sequential plugin registration at startup.
var (
    globalRegistry     *PluginRegistry
    globalRegistryOnce sync.Once
)

// GetGlobalRegistry returns the global plugin registry instance,
// initializing it on first access using sync.Once for thread safety.
// This function is safe to call concurrently from multiple goroutines.
func GetGlobalRegistry() *PluginRegistry {
    globalRegistryOnce.Do(func() {
        globalRegistry = NewPluginRegistry()
    })
    return globalRegistry
}
```

**Option B: Defensive Synchronization (Not Recommended)**

Adding an extra `sync.RWMutex` would be redundant given Go's memory model guarantees and would add unnecessary overhead:

```go
// Not recommended - sync.Once already provides necessary guarantees
var globalRegistryMu sync.RWMutex

func GetGlobalRegistry() *PluginRegistry {
    globalRegistryOnce.Do(func() {
        globalRegistry = NewPluginRegistry()
    })
    globalRegistryMu.RLock()
    defer globalRegistryMu.RUnlock()
    return globalRegistry
}
```

## Acceptance Criteria

- [ ] Add comprehensive documentation to `globalRegistry` variable explaining thread-safety guarantees
- [ ] Document the initialization-before-concurrent-access usage pattern
- [ ] Add documentation to `GetGlobalRegistry()` confirming concurrent-safety
- [ ] Verify `go test -race` continues to pass
- [ ] Review and document the plugin lifecycle in `InitializePlugins()`

## Additional Context

**Current Plugin Initialization Flow:**
1. `InitializePlugins()` is called during application startup
2. Plugins are registered sequentially via `RegisterGlobalPlugin()`
3. After initialization, `GetGlobalRegistry()` may be called concurrently during request handling

**Related Files:**
- `internal/audit/plugin.go` - Global registry implementation
- `internal/compliance/plugin.go` - Plugin interface definition

**References:**
- [Go Memory Model - sync.Once](https://go.dev/ref/mem#once)
- [Effective Go - Initialization](https://go.dev/doc/effective_go#init)

### PR DESCRIPTION
**Pull request by unclesp1d3r**
## Summary

Document thread-safety guarantees for the global `PluginRegistry` singleton pattern and apply Go modernize fixes across the codebase.

**Impact**: 24 files changed (+188, -134 net relevant lines) | **Risk Level**: Low | **Review Time**: ~10 minutes

Closes #290

---

### Documentation: Thread-Safety Guarantees (core issue)

- **`internal/audit/plugin.go`** -- Expanded godoc for `globalRegistry`, `GetGlobalRegistry()`, and `RegisterGlobalPlugin()` with:
  - `sync.Once` memory model citation ([go.dev/ref/mem#once](https://go.dev/ref/mem#once)) and happens-before explanation
  - Lifecycle contract: register-at-startup, read-during-operation
  - Explicit note that `PluginManager` maintains an independent registry
- **`internal/audit/plugin_manager.go`** -- Expanded godoc for `InitializePlugins()` documenting sequential-only initialization and registry independence
- **`AGENTS.md`** section 8.1 -- Added `PluginManager` vs global registry independence note
- **`GOTCHAS.md`** (new) -- Developer-facing pitfalls document covering registry independence, `t.Parallel()` with global state, map iteration order, and XML presence detection
- **`CONTRIBUTING.md`** -- Added GOTCHAS.md cross-references and fixed email link syntax

### Code: Go Modernize Fixes

| Pattern | Before | After | Files |
|---------|--------|-------|-------|
| Sort functions | `sort.Strings()` / `sort.Slice()` | `slices.Sort()` / `slices.SortFunc()` | 9 files |
| Interface type | `interface{}` | `any` | 3 files |
| Range over int | `for i := 0; i < n; i++` | `for i := range n` | 1 file |
| Dead code removal | Hand-rolled `extractOctets()` | `strings.Split()` | 1 file |

**Affected packages**: `cmd/`, `internal/converter/formatters/`, `internal/diff/`, `internal/docgen/`, `internal/processor/`, `internal/sanitizer/`, `internal/schema/`, `internal/testing/`, `internal/`, `tools/docgen/`

### Docs and Config Fixes

- **`docs/dev-guide/api.md`** -- Escaped `*ComplianceSummary` pointer notation for correct markdown rendering
- **`AGENTS.md`** section 5.10 -- Added `modernize` linter row for `sort.*` to `slices.*` migration
- **`AGENTS.md`** section 5.11 -- Updated sort function recommendations (removed deprecated `sort.Strings()`)

---

## Why These Changes

1. **Thread-safety documentation** (issue #290): The global `PluginRegistry` uses `sync.Once` + `sync.RWMutex` correctly, but the thread-safety contract and the two-registry distinction (`PluginManager` vs global singleton) were undocumented, creating a contributor trap
2. **Go modernize**: The `sort` package functions are flagged by the `modernize` linter; `slices.Sort`/`slices.SortFunc` are the project standard per AGENTS.md
3. **Dead code removal**: `extractOctets()` reimplemented `strings.Split()` -- removed per AGENTS.md section 5.14
4. **GOTCHAS.md**: Centralizes hard-won lessons that were previously only in agent memory or scattered across commit messages

---

## Type of Change

- [x] Documentation improvement
- [x] Code modernization (no behavior change)
- [x] Dead code removal
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change

---

## How Has This Been Tested?

- [x] `just ci-check` passes locally (pre-commit hooks, lint 0 issues, all 32 test packages green)
- [x] All CI checks pass on push (Lint, Build, Tests on ubuntu/macOS/Windows, Integration Tests, CodeQL, Coverage, FOSSA, Performance Benchmarks, Startup Time)
- [x] Go code review (go-reviewer): no critical/high findings
- [x] CodeRabbit CLI review: no findings

### Behavioral Note

`extractOctets` and `strings.Split` differ on edge cases (empty string, trailing dot), but `MapPrivateIP` is only called with validated IP addresses, and the `minOctetsForStructure` guard handles malformed input. The new `TestMapPrivateIP_ShortIP` test covers the fallback path.

---

### General
- [x] Code follows project style guidelines (AGENTS.md)
- [x] Self-review completed
- [x] No debugging code left
- [x] No sensitive data exposed

### Documentation
- [x] Thread-safety contract documented with Go memory model citations
- [x] AGENTS.md updated to reflect changes
- [x] GOTCHAS.md added and cross-referenced from CONTRIBUTING.md
- [x] Linter table updated with new `modernize` pattern

### Code Quality
- [x] All `sort.*` calls migrated to `slices.*` with `strings.Compare`
- [x] Dead code (`extractOctets`) removed, not suppressed
- [x] Tests updated to match new behavior
- [x] No breaking changes to public APIs

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -2,11 +2,14 @@
 
 This document consolidates all development standards, architectural principles, and workflows for the opnDossier project.
 
+@GOTCHAS.md
+
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply AGENTS.md]: Don't care. It serves the purpose. This is for AI, not humans.
